<?xml version="1.0"?>
<launch>
    <!-- arguments -->
    <arg name="paused" default="false"/>
    <arg name="use_sim_time" default="true"/>
    <arg name="gui" default="false"/>
    <arg name="headless" default="false"/>
    <arg name="debug" default="false"/>
    <arg name="start_moveit" default="false"/>
    <arg name="load_gripper" default="false"/>

    <!-- Load robot description from xacro file -->
    <param name="robot_description" command="$(find xacro)/xacro --inorder $(find panda_table_description)/urdf/panda_table.urdf.xacro load_gripper:=$(arg load_gripper)"/>

    <!-- Launch GAZEBO with own world configuration -->
    <include file="$(find gazebo_ros)/launch/empty_world.launch">
        <arg name="debug" value="$(arg debug)"/>
        <arg name="gui" value="$(arg gui)"/>
        <arg name="paused" value="$(arg paused)"/>
        <arg name="use_sim_time" value="$(arg use_sim_time)"/>
        <arg name="headless" value="$(arg headless)"/>
    </include>

    <node name="spawn_urdf" pkg="gazebo_ros" type="spawn_model" args="-param robot_description -urdf -model panda -J panda_joint1 0.0 -J panda_joint2 -0.8 -J panda_joint3 0.0 -J panda_joint4 -2.8 -J panda_joint5 0.0 -J panda_joint6 1.8 -J panda_joint7 0.8"/>

    <!-- Load joint controller configurations from YAML file to parameter server -->
    <rosparam file="$(find panda_table_description)/config/panda_gazebo_control.yaml" command="load"/>
    <rosparam file="$(find force_controllers)/config/panda_impedance.yaml" command="load"/>
    
    <rosparam command="load" file="$(find panda_insertion)/config/spring_parameters_sim.yaml" />

    <!-- Load the controllers -->
    <node name="controller_spawner" pkg="controller_manager" type="spawner" respawn="false" output="screen" args="joint_state_controller effort_joint_trajectory_controller" /> 

    <!-- Convert joint states to TF transforms for rviz, etc -->
    <node name="robot_state_publisher" pkg="robot_state_publisher" type="robot_state_publisher" respawn="true" output="screen" />

    <!-- Custom console config for dubugging messages -->
    <env name="ROSCONSOLE_CONFIG_FILE" value="$(find panda_insertion)/config/custom_rosconsole.conf"/>

    <!-- Nodes -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find panda_table_description)/launch/sim.rviz" />

    <node name="insertion" pkg="panda_insertion" type="insertion" output="screen">
        <param name="jointTrajectoryTopic" type="str" value="effort_joint_trajectory_controller/command"/>
        <param name="equilibriumPoseTopic" type="str" value="impedance_controller/equilibrium_pose"/>
      <param name="desiredStiffnessTopic" type="str" value="impedance_controller/desired_stiffness"/>

        <param name="loadControllerService" type="str" value="controller_manager/load_controller"/>
        <param name="switchControllerService" type="str" value="controller_manager/switch_controller"/>

        <param name="impedanceController" type="str" value="impedance_controller"/>
        <param name="positionJointTrajectoryController" type="str" value="effort_joint_trajectory_controller"/>

	<remap from="/panda/franka_state_controller/F_ext" to="/force"/>

        <param name="x_goal" type="double" value="0.475"/>
        <param name="y_goal" type="double" value="0.105"/>
        <param name="z_goal" type="double" value="0.74"/>
        <!-- <param name="x_goal" type="double" value="0.0"/>
        <param name="y_goal" type="double" value="0.0"/>
        <param name="z_goal" type="double" value="0.9"/> -->
    </node>
</launch>
